presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotBrokerPageSourceV2.java

```diff
@@ -0,0 +1,129 @@
+    }
+
+    @Override
+    protected GeneratedPinotQuery getBrokerQuery()
+    {
+        return brokerSql;
+    }
+
+    @Override
+    String getQueryUrlTemplate()
+    {
+        return QUERY_URL_TEMPLATE;
+    }
+
+    @Override
+    String getRequestPayloadTemplate()
+    {
+        return REQUEST_PAYLOAD_TEMPLATE;
+    }
+
+    @VisibleForTesting
+    @Override
+    public int populateFromQueryResults(
+            GeneratedPinotQuery pinotQuery,
+            List<BlockBuilder> blockBuilders,
+            List<Type> types,
+            String body)
+    {
+        String sql = pinotQuery.getSql();
+        JsonNode jsonBody;
+        try {
+            jsonBody = objectMapper.readTree(body);
+        }
+        catch (IOException e) {
+            throw new PinotException(PINOT_UNEXPECTED_RESPONSE, Optional.of(sql), "Couldn't parse response", e);
+        }
+        handleCommonResponse(sql, jsonBody);
+        JsonNode resultTable = jsonBody.get("resultTable");
+        if (resultTable != null) {
+            JsonNode dataSchema = resultTable.get("dataSchema");
+            if (dataSchema == null) {
+                throw new PinotException(
+                    PINOT_UNEXPECTED_RESPONSE,
+                    Optional.of(sql),
+                    String.format("Expected data schema in the response"));
+            }
+            JsonNode columnDataTypes = dataSchema.get("columnDataTypes");
+            JsonNode columnNames = dataSchema.get("columnNames");
+
+            if (columnDataTypes == null
+                    || !columnDataTypes.isArray()
+                    || columnDataTypes.size() != blockBuilders.size()) {
+                throw new PinotException(
+                    PINOT_UNEXPECTED_RESPONSE,
+                    Optional.of(sql),
+                    String.format("ColumnDataTypes and results expected for %s, expected %d columnDataTypes but got %d", sql, blockBuilders.size(), columnDataTypes == null ? 0 : columnDataTypes.size()));
+            }
+            if (columnNames == null
+                    || !columnNames.isArray()
+                    || columnNames.size() != blockBuilders.size()) {
+                throw new PinotException(
+                    PINOT_UNEXPECTED_RESPONSE,
+                    Optional.of(sql),
+                    String.format("ColumnNames and results expected for %s, expected %d columnNames but got %d", sql, blockBuilders.size(), columnNames == null ? 0 : columnNames.size()));
+            }
+
+            JsonNode rows = resultTable.get("rows");
+            checkState(rows.size() >= 1, "Expected at least one row to be present");
+            setRows(sql, blockBuilders, types, rows);
+            return rows.size();
+        }
+        else {
+            throw new PinotException(
+                    PINOT_UNEXPECTED_RESPONSE,
+                    Optional.of(sql),
+                    "Expected one of aggregationResults or selectionResults to be present");
```

Comment:
Error message should state that resultTable is missing.

Reply:
will do